### ISSUE TITLE
APK with coroutines can't be installed on some Android 5 devices

### ISSUE THREAD

**Issue by kzotin** [Kotlin/kotlinx.coroutines#1979]
TL;DR: even empty project with coroutines can't be installed on **Xiaomi Redmi Note 2** and some other devices. This blocks us from using coroutines and **androidx-ktx** libraries (even without explicit coroutines usage)

Hi,

We're having quite a popular Android App (1.5M MAU, 0.5M DAU), and with a recent update, some android devices can't update/install the App, including:
 
- **Xiaomi Redmi Note 2 (Helio)**
- **Prestigio PSP3504DUO**
- **Lenovo A5000 5.0.2(21)**  and others

Finding root cause comes down to APK produced by D8 when coroutines are there. Even empty project with `kotlinx-coroutines-core` cant be installed with the following error:

```
Installation did not succeed.
The application could not be installed: INSTALL_FAILED_DEXOPT
The device might have stale dexed jars that don't match the current version (dexopt error)
```

Sample project with correct & incorrect APK can be found here:
https://github.com/kzotin/deployerror-ktx

Downgrading coroutines showed that the issue appeared with 1.2.1 -> 1.2.2 update (and stays in 1.3.5). 

⚠️ Also even if coroutines are not used, they may be added by almost any `androidx-ktx` library. So having only `androidx.fragment:fragment-ktx:1.2.1` breaks installation, but `androidx.fragment:fragment:1.2.1` works fine.

Issue might be related to following one - it was a blocker for us using coroutines a year ago 😞 
https://github.com/Kotlin/kotlinx.coroutines/issues/925#issuecomment-467499537

Also turning off D8 fixes the problem, but not really an option - it's official compiler, forced by latest tools.

Based on the issue in 1st comment, the issue comes down to `kotlinx.coroutines.ChannelsKt__Channels_commonKt.filterNotNullTo`. 
Updating to the latest AGP 4.0 beta or 4.1 alpha did not change anything. 

I've got Xiaomi Redmi Note 2 device, where issue is reproduced. Please let me know if I can assist somehow.

Thanks in advance!

**Comment by kzotin** [comment]
Related issue: https://issuetracker.google.com/issues/149770839

**Comment by Tolriq** [comment]
Could also be https://issuetracker.google.com/issues/126056766 unfortunately this is 100% an D8/R8 issue with some older devices and they do not seem to have those kind of issues as high priority.

Nothing to be done on coroutine side.

**Comment by elizarov** [comment]
A general comment here: We might be able to do something on `kotlinx.coroutines` library side to avoid such problems, but we can do it if and only if we have a fully automated test that reproduces this issue and fails with the current version of code we have. If someone contributes such a test, then we can figure out what changed on 1.2.1->1.2.2 transition and we might be able to rewrite code somehow to work around this problem. However, without a test, there's nothing we can really do.

**Comment by Tolriq** [comment]
@elizarov this can not be reproduced by a test and the versions of coroutines generating this depends on the devices / cases see https://issuetracker.google.com/issues/126056766#comment19 this one is not about 1.2.1 -> 1.2.2.

The issue is the dex compilation by ART on specific devices with specific android versions, generated by some specific optimisations that D8 applies that modifies coroutines code this can not be exported to be reproduced elsewhere :(
This is a bug in those compilers and only those, the only reproduction possible is to compile the DEX file on those devices. If you have one of those devices then the reproduction is easy install any APK and see the compiler crash.

The only workaround on coroutines side would be to prevent D8 optimisation from happening but since you would not be able to target only those devices it would prevent those optimisation on the 99,99% of devices and so could have performance impact for the vast majority.
